File: android/app/src/main/java/com/touliao/app/feature/chat/OwnedTextSender.kt
package com.touliao.app.feature.chat

import com.touliao.app.core.storage.OutboxOwner
import com.touliao.app.core.storage.OutboxStore
import com.touliao.app.core.storage.TokenStore
import com.touliao.app.data.model.Message

/** The VM and tests share this actual send/ACK persistence path. */
internal suspend fun sendOwnedText(
    message: Message,
    owner: OutboxOwner,
    credential: TokenStore.Snapshot,
    tokens: TokenStore,
    outbox: OutboxStore,
    send: suspend (TokenStore.Snapshot) -> Result<Message>,
    onSuccess: (Message) -> Unit,
    onFailure: () -> Unit,
) {
    if (message.sender_id != owner.accountId || !tokens.isCurrent(credential)) return
    val result = send(credential)
    tokens.withCurrent(credential) {
        result.onSuccess { real ->
            if (real.sender_id != owner.accountId || real.conversation_id != message.conversation_id) return@onSuccess
            outbox.remove(message.conversation_id, message.id, owner)
            onSuccess(real)
        }.onFailure {
            outbox.upsert(message.conversation_id, message, owner)
            onFailure()
        }
    }
}

File: ios/Touliao/Features/Chat/OwnedTextSender.swift
import Foundation

/// Shared by the VM and isolated tests; transport is the only substituted boundary.
@MainActor
func sendOwnedText(message: Message, owner: OutboxOwner, credential: KeychainStore.Snapshot,
                   credentials: KeychainStore, outbox: OutboxStore,
                   send: (KeychainStore.Snapshot) async -> Result<Message, Error>,
                   onSuccess: (Message) -> Void, onFailure: () -> Void) async {
    guard message.senderId == owner.accountId, credentials.isCurrent(credential) else { return }
    let result = await send(credential)
    credentials.withCurrent(credential) {
        switch result {
        case .success(let real):
            guard real.senderId == owner.accountId, real.conversationId == message.conversationId else { return }
            outbox.remove(message.conversationId, message.id, owner: owner)
            onSuccess(real)
        case .failure:
            outbox.upsert(message.conversationId, message, owner: owner)
            onFailure()
        }
    }
}
